scheduler/supervisor/cdn_test.go

```diff
@@ -0,0 +1,504 @@
+				mockPeer.Host = mockHost
+				return patch
+			},
+			expect: func(t *testing.T, cdn CDN, peer *Peer, err error) {
+				assert := assert.New(t)
+				assert.Equal(mockPeer, peer)
+				assert.Nil(err)
+			},
+		},
+		{
+			name:   "downloadTinyFile success (restore host from client)",
+			status: TaskStatusWaiting,
+			mock: func(t *testing.T) *gomonkey.Patches {
+				patch := gomonkey.ApplyMethodSeq(reflect.TypeOf(mockLogger), "Debugf",
+					[]gomonkey.OutputCell{{Values: gomonkey.Params{}}})
+
+				patch.ApplyMethodSeq(reflect.TypeOf(mockTask), "Log",
+					[]gomonkey.OutputCell{{Values: gomonkey.Params{mockLogger}}})
+
+				const testwords string = "dragonfly-scheduler-test"
+				res := &http.Response{
+					Body: ioutil.NopCloser(
+						bytes.NewBuffer([]byte(testwords))),
+				}
+				httpRet := []gomonkey.OutputCell{
+					{Values: gomonkey.Params{res, nil}},
+				}
+				patch.ApplyFuncSeq(http.Get, httpRet)
+
+				patch.ApplyMethodSeq(reflect.TypeOf(mockPeer), "Touch",
+					[]gomonkey.OutputCell{{Values: gomonkey.Params{}}})
+
+				newPeerRet := []gomonkey.OutputCell{
+					{Values: gomonkey.Params{mockPeer}},
+				}
+				patch.ApplyFuncSeq(NewPeer, newPeerRet)
+
+				hostFromCDNRet := []gomonkey.OutputCell{
+					{Values: gomonkey.Params{mockHost, true}},
+				}
+				patch.ApplyMethodSeq(reflect.TypeOf(mockCDNDynmaicClient), "GetHost", hostFromCDNRet)
+
+				hostRet := []gomonkey.OutputCell{
+					{Values: gomonkey.Params{nil, false}},
+				}
+				patch.ApplyMethodSeq(reflect.TypeOf(mockHostManager), "Get", hostRet)
+				patch.ApplyMethodSeq(reflect.TypeOf(mockHostManager), "Add",
+					[]gomonkey.OutputCell{{Values: gomonkey.Params{}}})
+
+				peerRet := []gomonkey.OutputCell{
+					{Values: gomonkey.Params{nil, false}},
+				}
+				patch.ApplyMethodSeq(reflect.TypeOf(mockPeerManager), "Get", peerRet)
+				patch.ApplyMethodSeq(reflect.TypeOf(mockPeerManager), "Add",
+					[]gomonkey.OutputCell{{Values: gomonkey.Params{}}})
+
+				streamRet := []gomonkey.OutputCell{
+					{Values: gomonkey.Params{mockPieceSeed, nil}},
+				}
+				patch.ApplyMethodSeq(reflect.TypeOf(mockPieceSeedStream), "Recv", streamRet)
+				clientRet := []gomonkey.OutputCell{
+					{Values: gomonkey.Params{mockPieceSeedStream, nil}},
+				}
+				patch.ApplyMethodSeq(reflect.TypeOf(mockCDNDynmaicClient), "ObtainSeeds", clientRet)
+
+				mockPieceSeed.Done = true
+				mockPieceSeed.ContentLength = int64(len(testwords))
+				mockHost.IP = "0.0.0.0"
+				mockHost.DownloadPort = 1919
+				mockTask.ID = "1919810"
+				mockPeer.Host = mockHost
+				return patch
+			},
+			expect: func(t *testing.T, cdn CDN, peer *Peer, err error) {
+				assert := assert.New(t)
+				assert.Equal(mockPeer, peer)
+				assert.Nil(err)
+				assert.Equal([]byte("dragonfly-scheduler-test"), mockTask.DirectPiece)
+			},
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			patch := tc.mock(t)
+			cdn := NewCDN(mockCDNDynmaicClient, mockPeerManager, mockHostManager)
```

Comment:
`mockCDNDynmaicClient`, `mockPeerManager` and `mockHostManager` used gomock.

Reply:
used in PeerManager and TaskManager initial.